### DISCUSSION TITLE
How to parse all outputs from template string?

### DISCUSSION THREAD

**Discussion by timsofteng** [harttle/liquidjs#528] [General]
Hello.
I need to parse all outputs from template.

```
{{bucket}}
{% for product in products %}
   {{ product.name }} ads
{% endfor %}
```
So I want to get array like `["bucket", "product.name"]` or something like that.
How can I do it?

**Comment by harttle** [comment]
There is a way, but now there's no AST definition, that means the way to access the parsed tokens is implementation specific and subject to change. For example the following code outputs what you mentioned above:

```javascript
const { Liquid } = require('liquidjs')
const engine = new Liquid()
const tpls = engine.parse(`
{{bucket}}
{% for product in products %}
   {{ product.name }} ads
{% endfor %}
`)
console.log(tpls[1].token.content) // outputs: bucket
console.log(tpls[3].impl.templates[1].token.content) // outputs: product.name
```

Note that the name `.impl` is defined in `for` implementation, and the way to access children nodes can be different for each tag. And `tpls[0]` is represents `\n` before the first output `{{bucket}}`.

I'm planning to provide a well-defined AST in the next major version (latter this year). Just created this work item to remind me: https://github.com/harttle/liquidjs/projects/1#card-84455385

**Comment by timsofteng** [comment]
Hm. I still don't understand how can I know specific indexes of array's elements? Imagine I have big complex template. How can I grab outputs?

**Reply by pbojinov** [reply]
My use case is similar, I wanted to find all liquid handles in a template. I solved it using the following approach.

Input:
```rb
My top selling product is {{ products[0].title }} with a total of {{ products[0].total_sales }}
```

Code:
```tsx
const template = engine.parse(cleanedHTML);

const handles = template
  .map(({token}: {token: Token}) => {
    if (token?.kind === TokenKind.Output) {
      return token.content;
    }
  })
  .filter((item) => item != null)

```

Output: 
```js
['products[0].title', 'products[0].total_sales']
```

**Comment by harttle** [comment]
In v10, I consolidated property names refering to child templates and created a demo for iterating templates. Sorry I forgot to update this thread after released v10. Here's the demo entry point to get all outputs: https://github.com/harttle/liquidjs/blob/master/demo/template/index.ts

The most important part is:
https://github.com/harttle/liquidjs/blob/7526d4062db7b15e735ac8cb9c1a9be0f86a347b/demo/template/get-outputs.ts#L8-L24

**Reply by harttle** [reply]
I created a PR to AST explorer, hopefully it'll be availabel soon https://github.com/fkling/astexplorer/pull/680
